src/util/netif.cpp

```diff
@@ -0,0 +1,287 @@
+        return std::nullopt;
+    }
+
+    // Get best route to default gateway.
+    // Leave interface_luid at all-zeros to use interface index instead.
+    status = GetBestRoute2(&interface_luid, best_if_idx, nullptr, &destination_address, 0, &best_route, &best_source_address);
+    if (status != NO_ERROR) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "Could not get best route for default route for interface index %d: %s\n",
+                best_if_idx, NetworkErrorString(status));
+        return std::nullopt;
+    }
+
+    Assume(best_route.NextHop.si_family == family);
+    if (family == AF_INET) {
+        return CNetAddr(best_route.NextHop.Ipv4.sin_addr);
+    } else if(family == AF_INET6) {
+        return CNetAddr(best_route.NextHop.Ipv6.sin6_addr, best_route.InterfaceIndex);
+    }
+    return std::nullopt;
+}
+
+#elif defined(__APPLE__)
+
+#include <net/route.h>
+#include <sys/sysctl.h>
+
+// Ensure correct alignment for sockaddrs.
+#define ROUNDUP(a) \
+    ((a) > 0 ? (1 + (((a) - 1) | (sizeof(uint32_t) - 1))) : sizeof(uint32_t))
+
+static std::optional<CNetAddr> FromSockAddr(const struct sockaddr* addr)
+{
+    // Fill in a CService from the sockaddr, then drop the port part.
+    CService service;
+    if (service.SetSockAddr(addr)) {
+        return (CNetAddr)service;
+    }
+    return std::nullopt;
+}
+
+static std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
+{
+    // MacOS: Get default gateway from route table.
+    // See man page for route(4) for the format.
+
+    // net.route.0.inet[6].flags.gateway
+    int mib[] = {CTL_NET, PF_ROUTE, 0, family, NET_RT_FLAGS, RTF_GATEWAY};
+    size_t l = 0;
+    if (sysctl(mib, sizeof(mib) / sizeof(int), 0, &l, 0, 0) < 0) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "Could not get sysctl length of routing table: %s\n", SysErrorString(errno));
+        return std::nullopt;
+    }
+    std::vector<std::byte> buf(l);
+    if (sysctl(mib, sizeof(mib) / sizeof(int), buf.data(), &l, 0, 0) < 0) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "Could not get sysctl data of routing table: %s\n", SysErrorString(errno));
+        return std::nullopt;
+    }
+    const struct rt_msghdr* rt = nullptr;
+    for (const std::byte* p = buf.data(); p < buf.data() + buf.size(); p += rt->rtm_msglen) {
+        // Iterate over routing entry addresses, get destination and gateway (if present).
+        rt = (const struct rt_msghdr*)p;
+        const struct sockaddr* sa = (const struct sockaddr*)(rt + 1);
+        std::optional<CNetAddr> dst;
+        std::optional<CNetAddr> gateway;
+        for (int i = 0; i < RTAX_MAX; i++) {
+            if (rt->rtm_addrs & (1 << i)) {
+                if (i == RTAX_DST) {
+                    dst = FromSockAddr(sa);
+                } else if (i == RTAX_GATEWAY) {
+                    gateway = FromSockAddr(sa);
+                }
+                sa = (const struct sockaddr*)((std::byte*)sa + ROUNDUP(sa->sa_len));
```

Comment:
This seems quite brittle and I don't fully understand it. I did test that it seems necessary, e.g. doing `sa++;` causes it to not map IPv6 ports. If I look at `struct sockaddr_storage` I'm seeing `_SS_ALIGNSIZE   (sizeof(__int64_t)`, but using `uint64_t` instead of `uint32_t` in your `ROUNDUP` also doesn't work.

Reply:
i'm not sure either, i see it in some other libraries, but no alignment is mentioned in the manual page for `route` at all. Leaving out `ROUNDUP` works fine btw. Might just delete it.

Reply:
> Also they're variable-length

That makes sense. Whereas `rt_msghdr` is fixed length I guess, so you're able to use `rt + 1` above.

Reply:
Right, let's add `sizeof(rt_msghdr)` manually instead of doing the +1 trick.

> That makes sense. Whereas rt_msghdr is fixed length I guess, so you're able to use rt + 1 above (I'm still a bit terified that compilers don't care).

If we can hardcode the assumption that `sa_len` is the first byte (which it is) then we can avoid casting and dereferencing sockaddr, anywhere.

Reply:
Currently i have the following, having changed the pointer arithmetic to byte offsets:
```c++
    // Iterate over messages (each message is a routing table entry).
    for (size_t msgptr = 0; msgptr < buf.size(); ) {
        const struct rt_msghdr* rt = (const struct rt_msghdr*)(buf.data() + msgptr);
        // Iterate over addresses within entry, get destination and gateway (if present).
        // Pointer to address data within message, starts after header.
        size_t saptr = msgptr + sizeof(rt_msghdr);
        std::optional<CNetAddr> dst;
        std::optional<CNetAddr> gateway;
        for (int i = 0; i < RTAX_MAX; i++) {
            if (rt->rtm_addrs & (1 << i)) {
                const struct sockaddr* sa = (const struct sockaddr*)(buf.data() + saptr);
                if (i == RTAX_DST) {
                    dst = FromSockAddr(sa);
                } else if (i == RTAX_GATEWAY) {
                    gateway = FromSockAddr(sa);
                }
                // Skip to next address.
                saptr += sa->sa_len;
            }
        }
        // Found default gateway?
        if (dst && gateway && dst->IsBindAny()) { // Route to 0.0.0.0 or :: ?
            return *gateway;
        }
        // Skip to next message.
        msgptr += rt->rtm_msglen;
    }
```

Reply:
That indeed looks better.

The *ptr variables are no longer actually pointers, so I suggest renaming them:

msgptr -> msg_pos
within entry - within message
saptr -> sa_pos
next_msgptr -> next_msg_pos

Let's set and check next_msgptr _before_ saptr and make it const.